src/metrics/include/metrics/v2/event.hpp

```diff
@@ -0,0 +1,282 @@
+        if (!status.ok()) {
+            BOOST_LOG_TRIVIAL(error) << "Couldn't close collector: " << status.error_message();
+        }
+    }
+
+
+private:
+    static auto createPath(const std::string& name, const std::string& path_prefix) {
+        std::stringstream str;
+        str << path_prefix << '/';
+        str << name << ".ftdc";
+        return str.str();
+    }
+
+    static poplar::CreateOptions createOptions(const std::string& name,
+                                               const std::string& path_prefix) {
+        poplar::CreateOptions options;
+        options.set_name(name);
+        options.set_events(poplar::CreateOptions_EventsCollectorType_BASIC);
+        options.set_path(createPath(name, path_prefix));
+        options.set_chunksize(1000);
+        options.set_streaming(true);
+        options.set_dynamic(false);
+        options.set_recorder(poplar::CreateOptions_RecorderType_PERF);
+        options.set_events(poplar::CreateOptions_EventsCollectorType_BASIC);
+        return options;
+    }
+
+    std::string _name;
+    poplar::PoplarID _id;
+    CollectorStubInterface _stub;
+};
+
+/**
+ * Primary point of interaction between v2 poplar internals and the metrics system.
+ */
+template <typename ClockSource, typename StreamInterface>
+class EventStream {
+    using duration = typename ClockSource::duration;
+    using OptionalPhaseNumber = std::optional<genny::PhaseNumber>;
+
+public:
+    explicit EventStream(const ActorId& actorId,
+                         const std::string& name,
+                         const OptionalPhaseNumber& phase,
+                         const std::string& path_prefix)
+        : _name{name}, _phase{phase}, _last_finish{ClockSource::now()} {
+        _metrics.set_name(_name);
+        _metrics.set_id(actorId);
+        this->_reset();
+    }
+
+    void addAt(const typename ClockSource::time_point& finish,
+               const OperationEventT<ClockSource>& event,
+               size_t workerCount) {
+        _metrics.mutable_time()->set_seconds(
+            std::chrono::time_point_cast<std::chrono::seconds>(finish).time_since_epoch().count());
+        _metrics.mutable_time()->set_nanos(
+            std::chrono::time_point_cast<std::chrono::nanoseconds>(finish)
+                .time_since_epoch()
+                .count());
+
+        _metrics.mutable_timers()->mutable_duration()->set_seconds(
+            event.duration.getSecondsCount());
+        _metrics.mutable_timers()->mutable_duration()->set_nanos(
+            event.duration.getNanosecondsCount());
+
+        _metrics.mutable_timers()->mutable_total()->set_seconds(
+            Period<ClockSource>(finish - _last_finish).getSecondsCount());
+        _metrics.mutable_timers()->mutable_total()->set_nanos(
+            Period<ClockSource>(finish - _last_finish).getNanosecondsCount());
+
+        _metrics.mutable_counters()->set_number(event.iters);
+        _metrics.mutable_counters()->set_ops(event.ops);
```

Comment:
elsewhere the `event` has names consistent with the poplar names - does `event.iters` want to be `event.number` just to keep the "glue" code as dumb as possible?